components/locale/src/langid.rs

```diff
@@ -0,0 +1,151 @@
+use std::fmt::Write;
+use std::str::FromStr;
+
+use crate::parser::{parse_language_identifier, ParserError, ParserMode};
+use crate::subtags;
+
+/// `LanguageIdentifier` is a core struct representing a [`Unicode Language Identifier`].
+///
+/// # Example
+///
+/// ```
+/// use icu_locale::LanguageIdentifier;
+///
+/// let li: LanguageIdentifier = "en-US".parse()
+///     .expect("Failed to parse.");
+///
+/// assert_eq!(li.language, "en");
+/// assert_eq!(li.script, None);
+/// assert_eq!(li.region.unwrap(), "US");
+/// assert_eq!(li.variants.len(), 0);
+/// ```
+///
+/// # Parsing
+///
+/// Unicode recognizes three levels of standard conformance for any language identifier:
+///
+///  * *well-formed* - syntactically correct
+///  * *valid* - well-formed and only uses registered language subtags, extensions, keywords, types...
```

Comment:
I don't think you care about extensions in language ids (that's part of locale class). Fix comment?